rfc-141-application-healthchecks.md

```diff
@@ -0,0 +1,224 @@
+
+If an app can't talk to a backing service it relies on, then it
+probably can't be trusted to handle any requests.  One of these
+healthchecks failing could indicate a networking or a configuration
+issue.
+
+Here are some examples of healthchecks which are unsuitable for load
+balancing or for post-deployment checks:
+
+- [`GovukHealthcheck::SidekiqQueueCheck`][], this is an abstract check
+  which other checks extend.
+  [`GovukHealthcheck::SidekiqQueueLatencyCheck`][] is an instance of
+  this check.
+- [`GovukHealthcheck::ThresholdCheck`][], this is another abstract
+  check which other checks extend.
+  [`GovukHealthcheck::SidekiqRetrySizeCheck`][] is an instance of this
+  check.
+- [`Healthcheck::ApiTokens`][] in [signon][], which reports that an
+  API token needs rotating.
+
+If one of these healthchecks fail, the app instances themselves are
+probably fine.  The problem is elsewhere.
+
+Two of these indicate capacity problems outside of the instance, and
+the third is an alert that a manual maintenance procedure needs to be
+performed some time in the next two months.  They all share the
+property that if one instance reports a failure, all will.  This makes
+them unsuitable for load balancing purposes.
+
+Even if we don't make our healthchecks suitable for load balancing, as
+GOV.UK moves towards continuous deployments, we will end up in
+situations where an automatic deployment is aborted because of
+something unrelated to the change being deployed.  That's not good.
+
+A healthcheck should check that the instance is running and isn't
+completely broken.  That's all.
+
+## Proposal
+
+My proposal comes down to three parts:
+
+1. Remove the "warning" state of `GovukHealthcheck`, so a healthcheck
+   result is either "ok" or "critical".
+2. Make `GovukHealthcheck` serve an HTTP status of 500 for a
+   "critical" result.
+3. Add a separate healthcheck for liveness, to use with the ECS
+   container lifecycle.
+
+But we can't make this change right now, we have a migration path to
+follow first.
+
+### Remove the "warning" state
+
+There are four possible semantics we could assign the "warning" state:
+
+| # | State    | Allows automatic deployments? | Allows requests to be sent to the instance? |
+| - | -------- | ----------------------------- | ------------------------------------------- |
+|   | ok       | yes | yes |
+| 1 | warning  | yes | yes |
+| 2 | warning  | yes | no  |
+| 3 | warning  | no  | yes |
+| 4 | warning  | no  | no  |
+|   | critical | no  | no  |
+
+- If we go for option 1, then "warning" is the same as "ok".
+- If we go for option 2, then "warning" will let us deploy unusable releases.
+- If we go for option 3, then "warning" will block deployments.
+- If we go for option 4, then "warning" is the same as "critical".
+
+The most sensible option is 3, which is our current behaviour.
+
+But does it really gain us anything over having separate alerts for
+the specific condition we need to know about?
```

Comment:
Is setting up separate alerts for these things part of this proposal?

If we just removed the warnings without implementing alerts, would that leave us with a problematic monitoring gap?

Reply:
It is, yes.  Step 2 of the migration path mentions making new alerts for checks which add value.

Reply:
Ah yep, gotcha. Sorry I missed that.